src/client.rs

```diff
@@ -22,88 +22,94 @@ use crate::Readdir;
 /// So it's better for us to not call `hdfsDisconnect` manually.
 /// Aka, don't implement `Drop` to disconnect the connection.
 ///
-/// Reference: [IOException: Filesystem closed exception when running oozie workflo](https://stackoverflow.com/questions/23779186/ioexception-filesystem-closed-exception-when-running-oozie-workflow)
+/// Reference: [IOException: Filesystem closed exception when running oozie workflow](https://stackoverflow.com/questions/23779186/ioexception-filesystem-closed-exception-when-running-oozie-workflow)
 ///
 /// # Examples
 ///
 /// ```
-/// use hdrs::Client;
+/// use hdrs::{Client, ClientBuilder};
 ///
-/// let fs = Client::connect("default");
+/// let fs = ClientBuilder::new("default".to_string()).with_user("default".to_string()).with_kerberos_ticket_cache_path("/tmp/krb5_111".to_string()).connect();
 /// ```
 #[derive(Debug)]
 pub struct Client {
     fs: hdfsFS,
 }
 
-/// HDFS's client handle is thread safe.
-unsafe impl Send for Client {}
-unsafe impl Sync for Client {}
+/// The builder of connecting to hdfs clusters.
+///
+/// # Examples
+///
+/// ```
+/// use hdrs::ClientBuilder;
+/// use hdrs::Client;
+///
+/// let fs = ClientBuilder::new("default".to_string()).with_user("default".to_string()).with_kerberos_ticket_cache_path("/tmp/krb5_111".to_string()).connect();
+/// ```
+pub struct ClientBuilder {
+    name_node: String,
+    user: Option<String>,
+    kerberos_ticket_cache_path: Option<String>
+}
 
-impl Client {
-    /// Connect to a name node with port
-    ///
-    /// Returns an [`io::Result`] if any error happens.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use hdrs::Client;
-    ///
-    /// let fs = Client::connect("default");
-    /// ```
-    pub fn connect(name_node: &str) -> io::Result<Self> {
-        prepare_env()?;
+impl<'a> ClientBuilder {
+    pub fn new(name_node: String) -> ClientBuilder {
+        ClientBuilder {
+            name_node,
+            user: None,
+            kerberos_ticket_cache_path: None
+        }
+    }
 
-        set_errno(Errno(0));
+    pub fn with_user(mut self, user: String) -> ClientBuilder {
```

Comment:
We should accept `&str` for the API easier to use. (This will apply to all APIs.)